pkg/controllers/ragengine_controller.go

```diff
@@ -33,14 +54,381 @@ func NewRAGEngineReconciler(client client.Client, scheme *runtime.Scheme, log lo
+						if featuregates.FeatureGates[consts.FeatureFlagKarpenter] {
+							if updateErr := c.updateStatusConditionIfNotMatch(ctx, ragEngineObj, kaitov1alpha1.ConditionTypeNodeClaimStatus, metav1.ConditionFalse,
+								"checkNodeClaimStatusFailed", err.Error()); updateErr != nil {
+								klog.ErrorS(updateErr, "failed to update ragengine status", "ragengine", klog.KObj(ragEngineObj))
+								return updateErr
+							}
+						} else {
+							if updateErr := c.updateStatusConditionIfNotMatch(ctx, ragEngineObj, kaitov1alpha1.ConditionTypeMachineStatus, metav1.ConditionFalse,
+								"checkMachineStatusFailed", err.Error()); updateErr != nil {
+								klog.ErrorS(updateErr, "failed to update ragengine status", "ragengine", klog.KObj(ragEngineObj))
+								return updateErr
+							}
+						}
+						return err
+					}
+				}
+				time.Sleep(1 * time.Second)
+			}
+			return nil
+		}
+	}
+}
+
 // SetupWithManager sets up the controller with the Manager.
 func (c *RAGEngineReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	c.Recorder = mgr.GetEventRecorderFor("RAGEngine")
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &corev1.Pod{},
+		"spec.nodeName", func(rawObj client.Object) []string {
+			pod := rawObj.(*corev1.Pod)
+			return []string{pod.Spec.NodeName}
+		}); err != nil {
+		return err
+	}
 	builder := ctrl.NewControllerManagedBy(mgr).
 		For(&kaitov1alpha1.RAGEngine{}).
+		Watches(&v1alpha5.Machine{}, c.watchMachines()).
 		WithOptions(controller.Options{MaxConcurrentReconciles: 5})
+	if featuregates.FeatureGates[consts.FeatureFlagKarpenter] {
+		builder.
+			Watches(&v1beta1.NodeClaim{}, c.watchNodeClaims()) // watches for nodeClaim with labels indicating ragengine name.
+	}
 	return builder.Complete(c)
 }
+
+// watches for machine with labels indicating RAGEngine name.
+func (c *RAGEngineReconciler) watchMachines() handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(
+		func(ctx context.Context, o client.Object) []reconcile.Request {
+			machineObj := o.(*v1alpha5.Machine)
+			name, ok := machineObj.Labels[kaitov1alpha1.LabelRAGEngineName]
+			if !ok {
+				return nil
+			}
+			namespace, ok := machineObj.Labels[kaitov1alpha1.LabelRAGEngineNamespace]
+			if !ok {
+				return nil
+			}
+			_, conditionFound := lo.Find(machineObj.GetConditions(), func(condition apis.Condition) bool {
+				return condition.Type == apis.ConditionReady &&
+					condition.Status == v1.ConditionTrue
+			})
+			if conditionFound && machineObj.DeletionTimestamp.IsZero() {
+				// No need to reconcile ragengine if the machine is in READY state unless machine is deleted.
+				return nil
+			}
+			return []reconcile.Request{
+				{
+					NamespacedName: client.ObjectKey{
+						Name:      name,
+						Namespace: namespace,
+					},
+				},
+			}
+		})
+}
+
+// watches for nodeClaim with labels indicating RAGEngine name.
+func (c *RAGEngineReconciler) watchNodeClaims() handler.EventHandler {
```

Comment:
same logic to https://github.com/Azure/kaito/blob/2ec45ac3141e955bcf37af78ba5f76f5f7e65cdd/pkg/controllers/workspace_controller.go#L922